File: src/FloodFill.java
class Solution {
 
      private int m, n, oldColor, newColor;
      private static int[][] directions = {                                        // directions matrix
          {-1, 0},
          {0, 1},
          {1, 0},
          {0, -1}
      };
  
      public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        
// ******************** BFS Approach ******************** 
  
        if(image[sr][sc] == color)                                               // if same color then return as it is
            return image;
        
        int m = image.length;
        int n = image[0].length;
        int oldColor = image[sr][sc];                                            // setting old color
        int newColor = color;                                                    // setting new color
        Queue<Pair<Integer, Integer>> q = new LinkedList<>();
        q.add(new Pair(sr, sc));

        while(!q.isEmpty()) {
            Pair<Integer, Integer> pair = q.poll();
            for(int[] dir : directions) {                                        // go for neighbours
                sr = pair.getKey();                                              // get row index from queue
                sc = pair.getValue();                                            // get column index from queue
                
                image[sr][sc] = newColor;                                        // set that grid value to newcolor
                
                int nr = sr + dir[0];
                int nc = sc + dir[1];
                if(nr<0 || nc<0 || nr==m || nc==n || image[nr][nc]!=oldColor)    // check bounds
                    continue;
                q.add(new Pair(nr, nc));                                         // adding indexes in queue
            }
        }
        
        return image;
  
// // ******************** DFS Approach ******************** 
        
//         if(image[sr][sc] == color)                                               // if same color then return as it is
//             return image;
        
//         m = image.length;
//         n = image[0].length;
//         oldColor = image[sr][sc];                                                // setting old color
//         newColor = color;                                                        // setting new color
        
//         dfs(image, sr, sc);                                                      // calling dfs
        
//         return image;

//     }
//     private void dfs(int[][] image, int sr, int sc) {
        
//         // base cae
//         if(sr<0 || sc<0 || sr==m || sc==n || image[sr][sc]!=oldColor)            // out of bounce OR not old color
//             return;
        
//         // logic
//         for(int[] dir : directions) {
//             image[sr][sc] = newColor;                                            // setting current grid value to new color
//             int nr = sr + dir[0];
//             int nc = sc + dir[1];
//             dfs(image, nr, nc);                                                  // calling dfs for neighbours
//         }

//     }
}

File: Flood_Fill.cpp
//BFS Approach
class Solution {
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        if(image.size()==0 )return image;
        int oldColor = image[sr][sc];
        if(image[sr][sc]==color)return image;
        int m=image.size(), n=image[0].size();
        queue<int> q;
        q.push(sr);
        q.push(sc);
        image[sr][sc]=color;
        int dir[4][2]={{0,1}, {0,-1}, {1,0}, {-1, 0}};
        while(!q.empty()){
            int cr=q.front();
            q.pop();
            int cc=q.front();
            q.pop();
            for(int i=0;i<4;i++){
                int nr = dir[i][0] + cr;
                int nc = dir[i][1] + cc;
                if(nr>=0 && nr< m && nc>=0 && nc<n && image[nr][nc]==oldColor){
                    q.push(nr);
                    q.push(nc);
                    image[nr][nc]=color;
                }

            }
        }
        return image;
    }
};
///////////////////////////////////////////////////////////////////////////////////////////////////////////
class Solution {
private:
    int dir[4][2]={{0,1}, {0,-1}, {1,0}, {-1,0}};
    int oldColor;

public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        if(image.size()==0 )return image;
        oldColor = image[sr][sc];
        if(image[sr][sc]==color)return image;
        int m=image.size(), n=image[0].size();
        dfs(image, sr, sc, color, m, n);
        return image;
    }
private:
    void dfs(vector<vector<int>>& image, int sr, int sc, int color, int m, int n){
        //base case
        if(sr<0 || sr==m || sc==n || sc<0 || image[sr][sc]!=oldColor){
            return;
        }

        //logic
        image[sr][sc]=color;
        for(int i=0;i<4;i++){
            int nr = dir[i][0] + sr; 
            int nc = dir[i][1] + sc;
            dfs(image, nr, nc, color, m, n);
        }

    }    
};
